docs/source/proposals/type-inference.rst

```diff
@@ -0,0 +1,121 @@
+======================
+NBEP ?: Type Inference
+======================
+
+:Author: Siu Kwan Lam
+:Date: Sept 2016
+:Status: Draft
+
+
+This document describes the current type inference implementation in numba.
+
+
+Introduction
+============
+
+Numba uses type information to ensure that every variable in the user code can
+be correctly lowered (translated into a low-level representation).  The type of
+a variable describes the set of valid operations and available attributes.
+Resolving this information during compilation avoids the overhead of type
+checking and dispatching at runtime.  However, Python is dynamically typed and
+the user does not declare variable types.  Since type information is absent,
+we use type inference to reconstruct the missing information.
+
+
+Numba Type Semantic
+===================
+
+Type inference operates on Numba IR, a mostly static-single-assignment (SSA)
+encoding of the Python bytecode.  Conceptually, all intermediate values in the
+Python code is explicitly assigned to a variable in the IR.  Numba enforces
+that each IR variable to have one type only.  A user variable (from the Python
+source code) can be mapped to multiple variables in the IR.  They are *versions*
+of a variable.  Each time a user variable is assigned to, a new version is
+created.  From that point, all subsequent references will use the new version.
+The user variable *evolves* as the function logic updates its type.  Merge
+points (e.g. subsequent block to an if-else, the loop body, etc..) in the control
+flow need extra care.  Each incoming paths could be holding a different
+variable version.  A new version is implicitly created at the merge point.
+The incoming versions are merged.  This may translate to an implicit cast.
+
+Numba uses function overloading to emulate Python duck-typing.  The type of a
+function can contain multiple call signatures that accept different argument
+types and yields different return types.  The process to decide the best
+signature for an overloaded function is called *overload resolution*.
+Numba partially implements the C++ overload resolution scheme
+(ISOCPP 13.3 Overload Resolution).  The scheme uses a "best fit" algorithm by
```

Comment:
The official standard is not free.  I can point to the working draft http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n4296.pdf